generate.go

```diff
@@ -0,0 +1,161 @@
+type GenerateFunc func(GenerateContext) (GenerateResult, error)
+
+// TODO: BuildContext provides the contextual details that are made available by the
+// buildpack lifecycle during the build phase. This context is populated by the
+// Build function and passed to BuildFunc during execution.
+type GenerateContext struct {
+	// Info includes the details of the buildpack parsed from the
+	// extension.toml included in the extension contents.
+	Info Info
+
+	// CNBPath is the absolute path location of the buildpack contents.
+	// This path is useful for finding the buildpack.toml or any other
+	// files included in the buildpack.
+	CNBPath string
+
+	// Platform includes the platform context according to the specification:
+	// https://github.com/buildpacks/spec/blob/main/buildpack.md#build
+	Platform Platform
+
+	// Plan includes the BuildpackPlan provided by the lifecycle as specified in
+	// the specification:
+	// https://github.com/buildpacks/spec/blob/main/buildpack.md#buildpack-plan-toml.
+	Plan BuildpackPlan
+
+	// Stack is the value of the chosen stack. This value is populated from the
+	// $CNB_STACK_ID environment variable.
+	Stack string
+
+	// WorkingDir is the location of the application source code as provided by
+	// the lifecycle.
+	WorkingDir string
+}
+
+// TODO: BuildResult allows buildpack authors to indicate the result of the build
+// phase for a given buildpack. This result, returned in a BuildFunc callback,
+// will be parsed and persisted by the Build function and returned to the
+// lifecycle at the end of the build phase execution.
+type GenerateResult struct {
+	// TODO: Config
+	ExtendConfig ExtendConfig
+
+	// TODO: BuildDockerfile
+	BuildDockerfile *Dockerfile
+	// TODO: RunDockerfile
+	RunDockerfile *Dockerfile
+}
+
+type ExtendConfig struct {
+	Build ExtendImageConfig `toml:"build"`
+}
+
+type ExtendImageConfig struct {
+	Args []ExtendImageConfigArg `toml:"args"`
+}
+
+type ExtendImageConfigArg struct {
+	Name  string `toml:"name"`
+	Value string `toml:"value"`
+}
+
+type Dockerfile string
+
+// TODO: Build is an implementation of the build phase according to the Cloud Native
+// Buildpacks specification. Calling this function with a BuildFunc will
+// perform the build phase process.
+func Generate(f GenerateFunc, options ...Option) {
+	config := OptionConfig{
+		exitHandler: internal.NewExitHandler(),
+		args:        os.Args,
+		tomlWriter:  internal.NewTOMLWriter(),
+		envWriter:   internal.NewEnvironmentWriter(),
+		fileWriter:  internal.NewFileWriter(),
+	}
+
+	for _, option := range options {
+		config = option(config)
+	}
+
+	pwd, err := os.Getwd()
+	if err != nil {
+		config.exitHandler.Error(err)
+		return
+	}
+
+	planPath := os.Getenv("CNB_BP_PLAN_PATH")
+
+	var plan BuildpackPlan
+	_, err = toml.DecodeFile(planPath, &plan)
+	if err != nil {
+		config.exitHandler.Error(err)
+		return
+	}
+
+	cnbPath := os.Getenv("CNB_EXTENSION_DIR")
+	outputPath := os.Getenv("CNB_OUTPUT_DIR")
```

Comment:
Maybe we should add a check here to ensure that the `outputPath` is definitely not set to the `<app>` directory

Reply:
What's the concern here? The `CNB_OUTPUT_DIR` env var comes from the lifecycle. I feel like we trust what the lifecycle passes to the extension.